### ISSUE TITLE
Make nightly.forgerock.financial work and pass functional tests

### ISSUE THREAD

**Issue by jorgesanchezperez** [SecureApiGateway/SecureApiGateway#304]
### Components

Identity-platform-initialiser, Dev Operations

### Describe the issue

Currently the initialiser populate the SBAT configuration to CDK deployment and FiDC (cloud platform), but only populate the configuration when the object of settings aren't created yet, the initialiser isn't able to override the configuration and the consequence of that is:
- The SBAT configuration on FiDC platform is out of date

In CDK is ok because is reset every day and always the configuration is up to date with the latest version of `master` branch.

### User Story (optional)

As the developer, I want to have a strategy to reset FiDC or override the configuration so that the initialiser populate the latest configuration changes and CDK and FiDC will be up to date always with the same configuration.

### Acceptance criteria (optional)

- [ ] CDK and FiDC always up to date with the configuration populated by the initialiser.
- [ ] CDK and FIDC with the same configuration always
### Related issues
- https://github.com/SecureBankingAccessToolkit/SecureBankingAccessToolkit/issues/301

### Approach solution
- The initialiser can check if the object configuration exist needs to update it.
- If the object configuration doesn't exist, create it
- Configuration flag to set true/false the update action

**Comment by dbadham-fr** [comment]
@jorgesanchezperez do we have any examples of running the initializer against FiDC failed? Do we see errors being raised when we attempt to write configuration that already exists?

Looking through the initializer code, we use a mixture of PUT and POST operations to create things. 
The PUT operations should do an update if the object already exists, we may need to set the `If-Match` header if this isn't working. See API docs: [REST Update](https://backstage.forgerock.com/docs/idcloud-idm/latest/crest/crest-update.html)
The POST operations will create new objects

**Comment by BohoCode** [comment]
We have issues in the nightly because the external secret manager that pulls secrets down from Google Secret Manager is defined in both these charts:

https://github.com/SecureApiGateway/securebanking-openbanking-uk-fidc-initializer/blob/e3416f6e8d57473f91ef1595d298e11e2ddb804a/_infra/helm/securebanking-openbanking-uk-iam-initializer/templates/cronjob-secret.yaml#L1

https://github.com/SecureApiGateway/securebanking-openbanking-test-data-initializer/blob/8eb73c694e4a0547cc3843a423cb9e4ffa9e746b/_infra/helm/securebanking-test-data-initializer/templates/cronjob-secret.yaml#L1

This means argo starts fighting to deploy them unless they are exactly the same in each repo. This has a bad smell. We should split this external secret manager helm stuff out into the new sapig-openbanking-uk-dev-envs repo and create an independent argo template for it.

**Comment by shaunharrisonFR** [comment]
Currently waiting on the initialiser go code to be modified to handle the 2fa bypass

**Comment by dbadham-fr** [comment]
The initializer can now authenticate and run successfully against FIDC.

This involved tweaking the auth flow to dismiss the 2FA popup.

An additional change was required to not configure the Platform Service in FIDC, as this is not permitted in the cloud. For CDK we are doing this to configure the cookie domain, an alternative solution for this is required for cloud.

**Comment by dbadham-fr** [comment]
## Issue with Password Grant OAuth2 flow (Resource Owner flow)

This is used to get an access_token for the ig service account to use to submit the DCR request to AM. The access token is required as the "Allow Open Dynamic Client Registration" is disabled; this is to enforce that DCR goes through IG in order to enforce OB/FAPI validations.

## Example logs (password redacted):
```
--- (request) id:d14d0cba-9858-4ddc-99cd-3d93484332c8-104 --->

POST https://openam-forgerock-securebankingaccelerato.forgeblocks.com/am/oauth2/realms/root/realms/alpha/access_token HTTP/1.1
Authorization: Basic aWctY2xpZW50OnBhc3N3b3Jk
Content-Length: 138
Content-Type: application/x-www-form-urlencoded

grant_type=password&username=service_account.ig&password=******%21&client_id=ig-client&scope=dynamic_client_registration+fr%3Aidm%3A*
2022-12-12 11:30:08.782 [vert.x-eventloop-thread-1] DEBUG io.netty.handler.ssl.SslHandler - [id: 0x9b544b59, L:/10.20.4.30:48282 - R:openam-forgerock-securebankingaccelerato.forgeblocks.com/35.190.21.35:443] HANDSHAKEN: protocol:TLSv1.2 cipher suite:TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
2022-12-12 11:30:08.818 [RxCachedThreadScheduler-3] INFO  o.f.o.d.c.C.c.ForgeRockClientHandler -

<--- (response) id:d14d0cba-9858-4ddc-99cd-3d93484332c8-104 ---

HTTP/2 500 Internal Server Error
alt-svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000
cache-control: no-store
content-length: 138
content-type: application/json; charset=UTF-8
date: Mon, 12 Dec 2022 11:30:08 GMT
pragma: no-cache
strict-transport-security: max-age=31536000; includeSubDomains; preload;
via: 1.1 google
x-content-type-options: nosniff
x-forgerock-transactionid: 1670844608795-9be57b214b100e799c44-969290
x-frame-options: SAMEORIGIN

{"error_description":"com.sun.identity.authentication.spi.AuthLoginException: No Configuration found|noConfig.jsp","error":"server_error"}
```
## Potential Fix
After some searching the following issues were found, the first issue is describing the same situation that we are facing, the second contains the fix on the FRAAS side:
https://bugster.forgerock.org/jira/browse/IAM-741
https://bugster.forgerock.org/jira/browse/FRAAS-4143

In short, we have had to fix the OAuth2Provider config to set `advancedOAuth2Config.passwordGrantAuthService` to  `PasswordGrant` (applied manually for now).

`PasswordGrant` exists as an Authentication Tree in the alpha realm in FIDC. This either came as part of the FIDC deployment or was added manually, there is nothing in the initializer to create this.

**NOTE** we are not configuring the passwordGrantAuthService for CDK environments, the configuration is not supplied by the initializer and has the value of `[Empty]` in the UI. Also, the `PasswordGrant` tree does not exist in the CDK env.

## Next steps
Investigate why this is needed for FIDC and not CDK

**Comment by dbadham-fr** [comment]
The next issue is with the hybrid flow to authorize the consent.

After the PSU logs in successfully, they are redirected to `/platform` instead of the expected destination: `https://obdemo.nightly.forgerock.financial/am/oauth2/realms/root/realms/alpha/authorize?client_id=fcccc74d-0821-4b39-984c-5a92a65ae17a&response_type=code%20id_token&redirect_uri=https://postman-echo.com/get&scope=openid%20payments&state=10d260bf-a7d9-444a-92d9-7b7a5f088208&nonce=10d260bf-a7d9-444a-92d9-7b7a5f088208&request=eyJ0eXAiOiJKV1QiLCJhbGciOiJQUzI1NiJ9.eyJpYXQiOjE2NzA4NTIxODgsIm5iZiI6MTY3MDg1MjE4OCwiZXhwIjoxNjcwODUyNzkzLCJqdGkiOiJKRlllT3R0ZjBmS0RfTTgwdGJGMDUiLCJhdWQiOiJodHRwczovL29iZGVtby5uaWdodGx5LmZvcmdlcm9jay5maW5hbmNpYWwvYW0vb2F1dGgyL3JlYWxtcy9yb290L3JlYWxtcy9hbHBoYSIsInNjb3BlIjoib3BlbmlkIGFjY291bnRzIiwiaXNzIjoiZmNjY2M3NGQtMDgyMS00YjM5LTk4NGMtNWE5MmE2NWFlMTdhIiwiY2xhaW1zIjp7ImlkX3Rva2VuIjp7ImFjciI6eyJ2YWx1ZSI6InVybjpvcGVuYmFua2luZzpwc2QyOmNhIiwiZXNzZW50aWFsIjp0cnVlfSwib3BlbmJhbmtpbmdfaW50ZW50X2lkIjp7InZhbHVlIjoiQUFDX2JjMzUzNmZmLWVjNTAtNDU5Ny1iMmI5LTk3ZWQwMDZjZTAxNyIsImVzc2VudGlhbCI6dHJ1ZX19fSwicmVzcG9uc2VfdHlwZSI6ImNvZGUgaWRfdG9rZW4iLCJyZWRpcmVjdF91cmkiOiJodHRwczovL3Bvc3RtYW4tZWNoby5jb20vZ2V0Iiwic3RhdGUiOiIxMGQyNjBiZi1hN2Q5LTQ0NGEtOTJkOS03YjdhNWYwODgyMDgiLCJub25jZSI6IjEwZDI2MGJmLWE3ZDktNDQ0YS05MmQ5LTdiN2E1ZjA4ODIwOCIsImNsaWVudF9pZCI6ImZjY2NjNzRkLTA4MjEtNGIzOS05ODRjLTVhOTJhNjVhZTE3YSJ9.gz8b2DAHgWWE5EqxYYkNN5haWkdi9caVyW8R_48ZyOJKMEmsO3PhYL8poMrH_Pozlz_RoobzEH9wE3Jgc6s8waZNoJf3mGVLeOsteaNEBDVVS2YlXGs7r-QQQYgE_a3UqYcMZSfsrhhUiB_stcTxTPOU25I70OzEfh82uujzhydqCEs6OqU447sgzKG95LXjCDCkbr5u8GYKxbEciwTZojaX-VRJC9iIHi1vS5a6eONfWYMAiRoUoHDkR5jsFdAqa3gGpZds7c7jpVJSzHM-EhNWGENqbaKAJraPOJRJtFFpRRxAnOoXTP-lYZUNe5jd_ID2CSNIKwNizIvoIUj-Jg&acr=urn%3Aopenbanking%3Apsd2%3Aca&acr_sig=1CV3mNNno7XVBxmVlhU3Q3_1Z5OdpshVpqtVIGcgg3U`

The Validation Service configures the acceptable redirect urls, this has been configured correctly in FIDC AM. Perhaps the validation rules applied in the cloud are more stringent i.e. must be on the same domain. I think it makes sense to configure the custom domain next and coming back to this issue afterwards.

**Comment by dbadham-fr** [comment]
Test data initializer has been fixed as per the iam initializer to authentication with FIDC:
See https://github.com/SecureApiGateway/securebanking-openbanking-test-data-initializer/pull/7

**Comment by dbadham-fr** [comment]
The functional tests are failing due to their being no configuration for the default login journey.

The authentication of the psu over rest is failing due to this.

The documentation suggests configuring a default authentcation service: https://backstage.forgerock.com/docs/idcloud-am/latest/authentication-guide/login-using-rest.html

This configuration has been applied to FIDC manually:
![Screenshot from 2022-12-14 13-09-56](https://user-images.githubusercontent.com/106093201/207604748-a65b75de-11c0-4b3c-a348-b978f52694e0.png)


To Do:

- [ ] Test if the above works for the psu login journey via the UI, can we remove the custom configuration mentioned in: https://github.com/SecureApiGateway/SecureApiGateway/issues/304#issuecomment-1346459777
- [ ] If the above is true then remove the additional OAuth2 provider configuration that was added
- [ ] Review if PasswordGrant is appropriate for this usage
- [ ] Automate configuring this via the iam initializer

**Comment by dbadham-fr** [comment]
Policy script now uses https and the IdentityPlatformFQDN to build URLs to call AM/IDM. This fixes the issue and works in all environments.

**Comment by dbadham-fr** [comment]
The functional tests were failing with a HTTP 502 Bad Gateway being returned by IG.
This is being returned as the call to AM /authorize is returning a response with header list size > 8k (the default max header list size configuration in Vertx). IG fails to decode the response with an exception.

```
Caused by: io.netty.handler.codec.http2.Http2Exception$HeaderListSizeException: Header size exceeded max allowed size (8192)
	at io.netty.handler.codec.http2.Http2Exception.headerListSizeError(Http2Exception.java:194)
	at io.netty.handler.codec.http2.Http2CodecUtil.headerListSizeExceeded(Http2CodecUtil.java:233)
	at io.netty.handler.codec.http2.HpackDecoder$Http2HeadersSink.finish(HpackDecoder.java:545)
	at io.netty.handler.codec.http2.HpackDecoder.decode(HpackDecoder.java:132)
	at io.netty.handler.codec.http2.DefaultHttp2HeadersDecoder.decodeHeaders(DefaultHttp2HeadersDecoder.java:126)
	at io.netty.handler.codec.http2.DefaultHttp2FrameReader$HeadersBlockBuilder.headers(DefaultHttp2FrameReader.java:743)
	at io.netty.handler.codec.http2.DefaultHttp2FrameReader$2.processFragment(DefaultHttp2FrameReader.java:483)
	at io.netty.handler.codec.http2.DefaultHttp2FrameReader.readHeadersFrame(DefaultHttp2FrameReader.java:491)
	at io.netty.handler.codec.http2.DefaultHttp2FrameReader.processPayloadState(DefaultHttp2FrameReader.java:254)
	at io.netty.handler.codec.http2.DefaultHttp2FrameReader.readFrame(DefaultHttp2FrameReader.java:160)
	at io.netty.handler.codec.http2.DefaultHttp2ConnectionDecoder.decodeFrame(DefaultHttp2ConnectionDecoder.java:181)
	at io.netty.handler.codec.http2.DecoratingHttp2ConnectionDecoder.decodeFrame(DecoratingHttp2ConnectionDecoder.java:63)
	at io.netty.handler.codec.http2.Http2ConnectionHandler$FrameDecoder.decode(Http2ConnectionHandler.java:378)
	... 34 common frames omitted
```

The fix is to increase the header list size to 16k via vertx configuration for our ReverseProxyHandler object configuration.

The response contains a large header list due to it needing to return the location header, which contains a request JWT and several other params, and 2 OAUTH_REQUEST_ATTRIBUTES cookies which contains the request JWT plus other request parameter data.
The cookie is returned twice as we need to use a custom domain for AM, so one cookie issued for the custom domain and one for the base domain.

**Comment by dbadham-fr** [comment]
With the above changes the functional tests are now passing against nightly.
